cunumeric/module.py

```diff
@@ -1330,6 +1332,131 @@ def append(arr, values, axis=None):
     return concatenate(inputs, axis)
 
 
+def block(arrays):
+    """
+    Assemble an nd-array from nested lists of blocks.
+
+    Blocks in the innermost lists are concatenated (see concatenate)
+    along the last dimension (-1), then these are concatenated along
+    the second-last dimension (-2), and so on until the outermost
+    list is reached.
+
+    Blocks can be of any dimension, but will not be broadcasted using
+    the normal rules. Instead, leading axes of size 1 are inserted,
+    to make block.ndim the same for all blocks. This is primarily useful
+    for working with scalars, and means that code like np.block([v, 1])
+    is valid, where v.ndim == 1.
+
+    When the nested list is two levels deep, this allows block matrices
+    to be constructed from their components.
+
+    Parameters
+    ----------
+    arrays : nested list of array_like or scalars (but not tuples)
+        If passed a single ndarray or scalar (a nested list of depth 0),
+        this is returned unmodified (and not copied).
+
+        Elements shapes must match along the appropriate axes (without
+        broadcasting), but leading 1s will be prepended to the shape as
+        necessary to make the dimensions match.
+
+    Returns
+    -------
+    block_array : ndarray
+        The array assembled from the given blocks.
+        The dimensionality of the output is equal to the greatest of: * the
+        dimensionality of all the inputs * the depth to which the input list
+        is nested
+
+    Raises
+    ------
+    ValueError
+        If list depths are mismatched - for instance, [[a, b], c] is
+        illegal, and should be spelt [[a, b], [c]]
+        If lists are empty - for instance, [[a, b], []]
+
+    See Also
+    --------
+    numpy.block
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+
+    """
+    # arrays should concatenate from innermost subarrays
+    # the 'arrays' should be balanced tree
+    active_arr = deque()
+    active_arr.append(((0,), arrays))
+    leaf_map = defaultdict(list)
+
+    # check if the 'arrays' is a balanced tree
+    while len(active_arr) > 0:
```

Comment:
I feel the whole implementation would be much more readable if it used recursion instead of a while loop unfolding each step at a time. Is it possible to reformulate this a bit?

Reply:
Yeah, this could be rewritten to a bottom-up recursion. I'll do it later after I do some analysis this week.

Reply:
Done!